### PR TITLE
Spike/result bindings

### PR DESCRIPTION
**Pull request by zonotope**
This patch refactors the query engine to maintain intermediate and final solutions to where searches as maps indexed by the relevant variables instead of vectors relying on positional ordering to associate variables with each item contained in the vector. 

This patch also builds the results of queries incrementally instead of all at once (unless there is an ordering or grouping specified in the query), which should limit memory consumption for queries with large result sets. Validation is centralized with clojure.spec, which should eventually provide a complete specification of FlureeQL allowing us to reuse that specification for validation across clients, servers, and on the web. Lastly, the operations that build the results of each query are more composable.

These changes should make the query engine more flexible because it should allow for query patterns to be reorderd without changing the results of the query execution or parsing, and that should make future opportunities for optimization of the query engine more apparent, and it should help make those optimizations also easier to implement.

I haven't implemented the "simple subject crawl" strategy with this query engine yet. That will come in a subsequent pull request. I also think there is a lot of room for cleanup and optimizations in the parsing code, but this pr is big enough as it is, so those changes too will come in another pr. 

The test suite passes on the clojure side, but there is a JavaScript failure that I haven't investigated yet, but I think the code changes are ready for review.

### CODE REVIEW

**Comment by dpetran** [comment]
I think you can fix the failing js tests by changing `fluree.db.query.fql.parse/parse-context` to do this conditional db-ctx:
```
(defn parse-context
  [q db]
  (let [db-ctx (get-in db [:schema (if (-> q :opts :js?)
                                     :context-str
                                     :context)])
        q-ctx  (or (:context q) (get q "@context"))]
    (json-ld/parse-context db-ctx q-ctx)))
```

I would like to explore expanding all incoming json-ld immediately once we get it so we don't have to pass context around internally, but that's a project for the future.

**Comment by zonotope** [comment]
> I think you can fix the failing js tests by changing `fluree.db.query.fql.parse/parse-context` to do this conditional db-ctx:
> 
> ```
> (defn parse-context
>   [q db]
>   (let [db-ctx (get-in db [:schema (if (-> q :opts :js?)
>                                      :context-str
>                                      :context)])
>         q-ctx  (or (:context q) (get q "@context"))]
>     (json-ld/parse-context db-ctx q-ctx)))
> ```
> 
> I would like to explore expanding all incoming json-ld immediately once we get it so we don't have to pass context around internally, but that's a project for the future.

Yep, that did it. Thanks!

**Comment by bplatz** [comment]
> I would like to explore expanding all incoming json-ld immediately once we get it so we don't have to pass context around internally, but that's a project for the future.

In addition to parsing the query itself, we use the context to produce the query results in the specified format. This latter part is the primary reason the context will need to be retained.

**Comment by zonotope** [comment]
closes #228. closes #276. closes #251

File: src/fluree/db/query/fql/syntax.cljc [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+(s/def ::depth nat-int?)
+
+(s/def ::prettyPrint boolean?)
+(s/def ::pretty-print ::prettyPrint)
+
+(s/def ::parseJSON boolean?)
+(s/def ::parse-json ::parseJSON)
+
+(s/def ::opts (s/keys :opt-un [::maxFuel ::max-fuel ::parseJSON ::parse-json
```
Comment (dpetran) [inline src/fluree/db/query/fql/syntax.cljc:22]: I love that this exists! How nice to see everything in one spot. One opt that is missing is the `::js?` key, which should be a boolean if it is present.

File: src/fluree/db/query/fql/parse.cljc [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+  [x]
+  (and (map? x)
+       (-> x keys first (= :filter))))
+
+(defn parse-filter-maps
+  [vars filters]
+  (let [vars (set vars)]
+    (->> filters
+         (mapcat vals)
+         flatten
```
Comment (dpetran) [inline src/fluree/db/query/fql/parse.cljc:268]: What shape do filters have? I'm curious what this `flatten` is clobbering.

File: src/fluree/db/query/fql/parse.cljc [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+         (map (fn [f-str]
+                (parse-filter-function f-str vars)))
+         (reduce (fn [m fltr]
+                   (let [var-name (::where/var fltr)]
+                     (update m var-name (fn [var-fltrs]
```
Comment (dpetran) [inline src/fluree/db/query/fql/parse.cljc:273]: This is elegant - I like how we just build up a sequence of filter functions per var.

File: src/fluree/db/query/fql/syntax.cljc [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                   ::depth ::opts ::prettyPrint ::pretty-print]))
+
+(defn validate
+  [qry]
+  (if (s/valid? ::analytical-query qry)
+    qry
+    (throw (ex-info "Invalid Query"
+                    {:status  400
+                     :error   :db/invalid-query
+                     :reasons (s/explain-data ::analytical-query qry)}))))
```
Comment (mpoffald) [inline src/fluree/db/query/fql/syntax.cljc:198]: 🤔  (Not a blocker for this PR, just an opinion) In the long-term I worry about throwing the output of `s/explain` out at people, I find them hard to parse even as a Clojure programmer (which not all users are going to be)

File: src/fluree/db/query/fql/parse.cljc [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+                     [v :asc]
+                     (let [[dir dim] ord
+                           v         (parse-var-name dim)]
+                       (if (syntax/asc? dir)
+                         [v :asc]
+                         [v :desc])))))))
+
+(defn parse-analytical-query
+  [q db]
+  (syntax/validate q)
```
Comment (mpoffald) [inline src/fluree/db/query/fql/parse.cljc:393]: I've never used spec as a validation step in production before, does/could this add any meaningful performance overhead?

File: src/fluree/db/query/fql/parse.cljc [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+            [fluree.db.query.analytical-filter :as filter]
+            [fluree.db.util.log :as log :include-macros true]
+            [fluree.db.dbproto :as dbproto]
+            [fluree.db.constants :as const]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn basic-query?
+  [q]
+  (contains? q :from))
```
Comment (mpoffald) [inline src/fluree/db/query/fql/parse.cljc:23]: So, looking at this and at the `syntax` ns, I'm still not sure what makes a basic query "basic." Is it just the use of the `from` key? Does that imply that there are no fancy aggregators, filters, etc, or something?

This is just a bit of internal terminology I've always been unclear about.

File: src/fluree/db/query/fql/parse.cljc [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+                       (get select-key)
+                       (parse-select-clause db context depth))]
+    (case select-key
+      :select     (assoc q :select select)
+      :select-one (assoc q :select-one select)
+      :selectOne  (-> q
+                      (dissoc :selectOne)
+                      (assoc :select-one select)))))
+
+(defn ensure-vector
```
Comment (dpetran) [inline src/fluree/db/query/fql/parse.cljc:364]: Could use the `sequential` util here as well.

File: src/fluree/db/query/fql/parse.cljc [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+    (let [iri     (iri->pred-id-strict p-iri db context)
+          recur-n (or recur-n util/max-integer)]
+      (where/->predicate iri recur-n))))
+
+(defn parse-full-text-predicate
+  [x db context]
+  (when (and (string? x)
+             (str/starts-with? x "fullText:"))
+    (-> x
+        (subs 9)
```
Comment (mpoffald) [inline src/fluree/db/query/fql/parse.cljc:220]: Looking through this PR, I've had this vague urge to suggest we add some tests for some of the helpers/internal representation steps. (Not as a blocker to this PR, necessarily, but ✨ someday✨ perhaps?) 

Parsing is implicitly tested by the query integration tests, but personally I'd like to have some tests for the pre-execution steps (validation, parsing). This new parsing pipeline is _much_ simpler than the old one, so the need for tests doesn't feel as urgent to me, but I still think it'd be valuable to document what this code is meant to output. 

While there's not _too_ much string-munging, regex, or magic number usage (this `9` isn't all that magical I guess), there is enough in here that it adds to my desire for tests. 😅 

Again, not a blocker for this PR, but something I've been thinking about.

File: src/fluree/db/query/exec.cljc [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  [db q]
+  (go
+   (let [error-ch  (async/chan)
+         result-ch (->> (where/search db q error-ch)
+                        (group/combine q)
+                        (order/arrange q)
+                        (drop-offset q)
+                        (take-limit q)
+                        (select/format db q error-ch)
+                        (collect-results q))]
```
Comment (mpoffald) [inline src/fluree/db/query/exec.cljc:57]: 😍 I love how easy it is to tell what's happening here.

Really the readability of this whole refactor is awesome, nice work

File: src/fluree/db/query/exec/where.cljc [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+    (reduce (fn [solution-ch pattern]
+              (with-constraint db pattern filters error-ch solution-ch))
+            initial-ch patterns)))
+
+(defmethod match-pattern :union
+  [db solution pattern _ error-ch]
+  (let [clauses   (val pattern)
+        clause-ch (async/to-chan! clauses)
+        out-ch    (async/chan 2)]
+    (async/pipeline-async 2
```
Comment (mpoffald) [inline src/fluree/db/query/exec/where.cljc:286]: I really love the use of `pipeline-async` wherever possible, it really makes it easier to reason about what's happening

File: src/fluree/db/query/fql/syntax.cljc (hunk 4 above)
Comment (zonotope) [reply src/fluree/db/query/fql/syntax.cljc:198]: Yep, I'm concerned by it too. The `explain-data` is just a temporary stop gap. I didn't want to let error messages hold up getting the query engine out, so I opened #312 to address this.

**Review by dpetran — APPROVED** [review]
:book:

File: src/fluree/db/query/exec/select.cljc [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  [solution db iri-cache compact error-ch selectors]
+  (if (sequential? selectors)
+    (go-loop [selectors  selectors
+              values     []]
+      (if-let [selector (first selectors)]
+        (let [value (<! (format-value selector db iri-cache compact error-ch solution))]
+          (recur (rest selectors)
+                 (conj values value)))
+        values))
+    (format-value selectors db iri-cache compact error-ch solution)))
```
Comment (mpoffald) [inline src/fluree/db/query/exec/select.cljc:113]: This is the world's smallest nitpick: I find it mildly stressful that the args for `format-value` and `format-values` are the same, but `selectors` and `solution` are swapped

File: src/fluree/db/query/fql/parse.cljc (hunk 2 above)
Comment (zonotope) [reply src/fluree/db/query/fql/parse.cljc:268]: each individual filter statement is a map with a single entry of the keyword `:filter` as key and a list (vector) of query functions defined as strings. The `vals` function gives a list of all the values, which is a list of a single element, which itself happens to be a list of filters. So, the mapcat combines the outer `vals` list into one, and the flatten combines the inner filter lists defined in each filter statement.

File: src/fluree/db/query/fql/parse.cljc (hunk 5 above)
Comment (zonotope) [reply src/fluree/db/query/fql/parse.cljc:393]: Yes, it could, but I'm not concerned at this point. There is definitely room for optimization in the current implementation as we throw away a lot of the work spec has done to decompose the query into components during validation only to go back and do the same decomposition during parsing. But, beyond that, I think we can gain a lot from normalizing and caching queries more than we do. If, after all that, spec still proves to be the bottleneck in query performance, and that proves to be a problem, it isn't that hard to pull out this validation path and go back to the old way. I think we can cross that bridge when we get to it.

File: src/fluree/db/query/fql/parse.cljc (hunk 6 above)
Comment (zonotope) [reply src/fluree/db/query/fql/parse.cljc:23]: I am also a little unclear, but, from what I could gather from reading the docs, looking at the code, and looking at examples, is that the basic query has a `:from` key while the analytical queries do not.

I think we should eventually treat them as two separate query languages with separate validation logic, because they are separate languages.

File: src/fluree/db/query/fql/parse.cljc (hunk 8 above)
Comment (zonotope) [reply src/fluree/db/query/fql/parse.cljc:220]: I agree. I think the internal parsed query representation is ready to be standardized and documented so that we can move to parsing SPARQL, SQL, and whatever other query languages we'd like to support into that representation directly instead of going through the analytical queries first, but I think that can wait a bit until we get all the FlureeQL stuff locked down.

File: src/fluree/db/query/exec/select.cljc (hunk 11 above)
Comment (zonotope) [reply src/fluree/db/query/exec/select.cljc:113]: It's not minor. I'll fix it.

File: src/fluree/db/query/fql/parse.cljc (hunk 6 above)
Comment (bplatz) [reply src/fluree/db/query/fql/parse.cljc:23]: We made the decision previously to deprecate basic queries. Basic queries (as it became unfortunately known as) mimicked a simple SQL query so were approachable as a first query for someone new to Fluree. It also made doing a query for a single subject super easy - something we said we could alternatively still support in a different way as per some ticket I can't find right now (but know it exists as I remember commenting on it).

Comment (bplatz) [reply src/fluree/db/query/fql/parse.cljc:23]: And, we want to also support the SPARQL `FROM` and `FROM NAMED` which uses `from` in a different way than basic queries had used them. https://www.w3.org/TR/rdf-sparql-query/#specifyingDataset

File: src/fluree/db/query/fql/parse.cljc (hunk 5 above)
Comment (bplatz) [reply src/fluree/db/query/fql/parse.cljc:393]: Probably worth coming to a decision on it however before writing code to turn s/explain output into good error messages. While it might have changed in the years since introduction, when spec was originally release it was recommended to not use it as an input validation and that was not a use case they were going to focus on supporting - which does introduce risk that future spec changes could further negatively impact this use case.

Comment (dpetran) [reply src/fluree/db/query/fql/parse.cljc:393]: Malli might be an alternative here - it's had a stable release and it's designed to be fast.

Comment (zonotope) [reply src/fluree/db/query/fql/parse.cljc:393]: Yeah, that's probably a better bet. @bplatz makes a good point about the aims of spec. Even if it's fast enough for us now, the fact that production performance isn't a goal of their development should still give us a lot of pause. Malli offers the same value prop and their development goals align more closely with our needs.

Comment (mpoffald) [reply src/fluree/db/query/fql/parse.cljc:393]: I second (third?) that, Malli would probably be a great choice here.

Thanks to having all the specs in one place, it shouldn't be too painful to change.
